src/core/Merger.cpp

```diff
@@ -0,0 +1,520 @@
+                           .arg(sourceEntry->title())
+                           .arg(sourceEntry->uuid().toHex());
+        } else if (timeTarget < timeSource) {
+            Entry* clonedEntry = sourceEntry->clone(Entry::CloneNewUuid | Entry::CloneIncludeHistory);
+            moveEntry(clonedEntry, context.m_targetGroup);
+            markOlderEntry(targetEntry);
+            changes << tr("Adding backup for older target %1 [%2]")
+                           .arg(targetEntry->title())
+                           .arg(targetEntry->uuid().toHex());
+        }
+        break;
+
+    case Group::KeepNewer:
+        if (timeTarget < timeSource) {
+            // only if other entry is newer, replace existing one
+            Entry* clonedEntry = sourceEntry->clone(Entry::CloneIncludeHistory);
+            Group* currentGroup = targetEntry->group();
+            qDebug("Updating entry %s.", qPrintable(targetEntry->title()));
+            moveEntry(clonedEntry, currentGroup);
+            eraseEntry(targetEntry);
+            changes << tr("Overwriting %1 [%2]").arg(clonedEntry->title()).arg(clonedEntry->uuid().toHex());
+        }
+        break;
+
+    case Group::KeepExisting:
+        break;
+
+    case Group::Synchronize:
+        if (timeTarget < timeSource) {
+            Group* currentGroup = targetEntry->group();
+            Entry* clonedEntry = sourceEntry->clone(Entry::CloneIncludeHistory);
+            qDebug("Merge %s/%s with alien on top under %s",
+                   qPrintable(targetEntry->title()),
+                   qPrintable(sourceEntry->title()),
+                   qPrintable(currentGroup->name()));
+            changes << tr("Synchronizing from newer source %1 [%2]")
+                           .arg(targetEntry->title())
+                           .arg(targetEntry->uuid().toHex());
+            moveEntry(clonedEntry, currentGroup);
+            mergeHistory(targetEntry, clonedEntry);
+            eraseEntry(targetEntry);
+        } else {
+            qDebug("Merge %s/%s with local on top/under %s",
+                   qPrintable(targetEntry->title()),
+                   qPrintable(sourceEntry->title()),
+                   qPrintable(targetEntry->group()->name()));
+            const bool changed = mergeHistory(sourceEntry, targetEntry);
+            if (changed) {
+                changes << tr("Synchronizing from older source %1 [%2]")
+                               .arg(targetEntry->title())
+                               .arg(targetEntry->uuid().toHex());
+            }
+        }
+        break;
+
+    default:
+        // do nothing
+        break;
+    }
+    return changes;
+}
+
+bool Merger::mergeHistory(const Entry* sourceEntry, Entry* targetEntry)
+{
+    const auto targetHistoryItems = targetEntry->historyItems();
+    const auto sourceHistoryItems = sourceEntry->historyItems();
+
+    QMap<QDateTime, Entry*> merged;
+    for (Entry* historyItem : targetHistoryItems) {
```

Comment:
That's the implemented strategy. It does look more complex since I do some assertions additionally to the needed boilerplate code to prevent unintended database changes.

Reply:
Why not simply use QSets then?

Reply:
`QSet` does not allow to specify arbitrary equality operators - it just uses `qHash`. Defining `qHash` using timestamps (and maybe uuid) only may lead to unintended behavior when defined globally (every `QSet` would treat entries equal regardless of object identity) or locally (different semantic of `QSet<Entry>`). In my opinion `QMap` states the intend more clearly and provides the needed characteristics.

Reply:
Okay, fair point. One last comment: you should perhaps merge lines 376 and 377 into an `else if ` construct to make it clearer that both conditions are mutually exclusive.